src/com/goide/editor/GoStatementMover.java

```diff
@@ -0,0 +1,143 @@
+      info.toMove2 = toMove2 == null ? null : new LineRange(toMove2);
+      return true;
+    }
+    return setUpInfo(info, elementRange, commonParent, down);
+  }
+
+  @Nullable
+  private static Couple<PsiElement> getLogicElementRange(@NotNull Couple<PsiElement> elementRange,
+                                                         @NotNull PsiElement parent) {
+    if (elementRange.first == null || elementRange.second == null) return null;
+    int start = elementRange.first.getTextOffset();
+    int end = elementRange.second.getTextRange().getEndOffset();
+
+    TextRange range = parent.getTextRange();
+    PsiElement[] children = parent.getChildren();
+    if (range.getStartOffset() == start && (children.length == 0 || children[0].getTextOffset() > start) ||
+        range.getEndOffset() == end && (children.length == 0 || children[children.length - 1].getTextRange().getEndOffset() < end)) {
+      return new Couple<PsiElement>(parent, parent);
+    }
+
+    PsiElement startElement = elementRange.first;
+    PsiElement endElement = elementRange.second;
+    for (PsiElement element : children) {
+      range = element.getTextRange();
+      if (range.getStartOffset() <= start && range.getEndOffset() <= end && range.getEndOffset() > start) {
+        startElement = element;
+      }
+      if (range.getStartOffset() >= start && range.getStartOffset() < end && range.getEndOffset() >= end) {
+        endElement = element;
+      }
+    }
+    if (!startElement.getParent().isEquivalentTo(endElement.getParent())) {
+      return null;
+    }
+
+    return new Couple<PsiElement>(startElement, endElement);
+  }
+
+  private static boolean setUpInfo(@NotNull MoveInfo info,
+                                   @NotNull Couple<PsiElement> range,
+                                   @NotNull PsiElement parent,
+                                   boolean down) {
+    info.toMove = new LineRange(range.first, range.second);
+    info.toMove2 = null;
+    GoTopLevelDeclaration block = PsiTreeUtil.getParentOfType(parent, GoTopLevelDeclaration.class);
+    if (block == null) return false;
+
+    int nearLine = down ? info.toMove.endLine : info.toMove.startLine - 1;
+    LineRange blockLineRange = new LineRange(block);
+    if (!blockLineRange.containsLine(down ? info.toMove.endLine + 1 : info.toMove.startLine - 2)) {
+      info.toMove2 = null;
+      return true;
+    }
+
+    info.toMove2 =
+      blockLineRange.containsLine(down ? info.toMove.endLine + 1 : info.toMove.startLine - 2)
+      ? new LineRange(nearLine, nearLine + 1)
+      : null;
+    return true;
+  }
+
+  @Nullable
+  private static PsiElement getElementOfTypeAfter(@NotNull PsiElement parent,
+                                                  @NotNull Couple<PsiElement> range,
+                                                  @NotNull Class<? extends PsiElement> clazz,
+                                                  boolean after) {
```

Comment:
Whose parent? What does `after` mean? How couple of element fit to `getElementOfTypeAfter` name?